File: src/arch_ops_server/boot.py
# SPDX-License-Identifier: GPL-3.0-only OR MIT
"""
Raspberry Pi bootloader management module.
Provides BOOT_ORDER inspection, permanent changes, and one-time next-boot device selection.
"""

import logging
import re
from pathlib import Path
from typing import Dict, Any, List, Optional

from .utils import (
    IS_ARCH,
    run_command,
    create_error_response,
    check_command_exists,
)

logger = logging.getLogger(__name__)

_STATE_FILE = Path("/var/lib/arch-ops-server/boot_order_restore")
_SERVICE_FILE = Path("/etc/systemd/system/arch-ops-boot-restore.service")

_BOOT_MODE_LABELS = {
    "0x0": "STOP",
    "0x1": "SD",
    "0x2": "NETWORK",
    "0x3": "RPIBOOT",
    "0x4": "USB",
    "0x5": "BCM-USB",
    "0x6": "NVMe",
    "0x7": "HTTP",
    "0xe": "STOP",
    "0xf": "RESTART",
}

_ORDER_PRESETS = {
    "sd_first": "0xf41",
    "nvme_first": "0xf46",
    "usb_first": "0xf14",
    "sd_nvme": "0xf16",
    "nvme_sd": "0xf61",
    "sd_only": "0xf1",
    "nvme_only": "0xf6",
    "usb_only": "0xf4",
}


def _decode_boot_order(boot_order: str) -> List[Dict[str, str]]:
    """Decode a BOOT_ORDER hex string into a list of human-readable modes."""
    hex_val = boot_order.lower().replace("0x", "")
    sequence = []
    for ch in reversed(hex_val):
        key = f"0x{ch}"
        sequence.append({
            "mode": key,
            "label": _BOOT_MODE_LABELS.get(key, f"UNKNOWN({key})"),
        })
    return sequence


async def _get_eeprom_config() -> Dict[str, str]:
    """Parse current EEPROM config into a dict. Returns empty dict on failure."""
    exit_code, stdout, _ = await run_command(
        ["rpi-eeprom-config"],
        timeout=10,
        check=False,
    )
    if exit_code != 0:
        return {}

    config = {}
    for line in stdout.strip().splitlines():
        line = line.strip()
        if "=" in line and not line.startswith("#"):
            key, _, val = line.partition("=")
            config[key.strip()] = val.strip()
    return config


async def _run_eeprom_update(key: str, value: str) -> Dict[str, Any]:
    """Update a single key in the EEPROM config via temp file."""
    exit_code, stdout, stderr = await run_command(
        ["rpi-eeprom-config"],
        timeout=10,
        check=False,
    )
    if exit_code != 0:
        return create_error_response("CommandError", f"Failed to read EEPROM: {stderr}")

    import tempfile
    with tempfile.NamedTemporaryFile(mode="w", suffix=".txt", delete=False) as tf:
        for line in stdout.strip().splitlines():
            line = line.strip()
            if line.startswith(f"{key}="):
                tf.write(f"{key}={value}\n")
            elif line and not line.startswith("#"):
                tf.write(f"{line}\n")
            elif line.startswith("#"):
                tf.write(f"{line}\n")
        tf.flush()
        tmp_path = tf.name

    exit_code, _, stderr = await run_command(
        ["rpi-eeprom-config", "--apply", tmp_path],
        timeout=15,
        check=False,
    )
    Path(tmp_path).unlink(missing_ok=True)
    if exit_code != 0:
        return create_error_response("CommandError", f"Failed to update EEPROM: {stderr}")
    return {"ok": True}


async def _detect_boot_devices() -> Dict[str, bool]:
    """Detect which boot devices are available (SD, NVMe, USB)."""
    devices = {"sd": False, "nvme": False, "usb": False}
    exit_code, stdout, _ = await run_command(
        ["lsblk", "-o", "NAME,TYPE,TRAN", "-n"],
        timeout=5,
        check=False,
    )
    if exit_code != 0:
        return devices

    for line in stdout.strip().splitlines():
        parts = line.strip().split()
        if len(parts) < 2:
            continue
        name = parts[0]
        if name.startswith("mmcblk"):
            devices["sd"] = True
        elif name.startswith("nvme"):
            devices["nvme"] = True
        elif name.startswith("sd"):
            devices["usb"] = True

    return devices


async def _install_restore_service(original_order: str) -> Dict[str, Any]:
    """Install a oneshot systemd service that restores BOOT_ORDER on next boot."""

    service_content = f"""[Unit]
Description=Restore RPi BOOT_ORDER after one-time boot
DefaultDependencies=no
After=local-fs.target
Before=basic.target

[Service]
Type=oneshot
ExecStart=/usr/bin/rpi-eeprom-config --apply BOOT_ORDER={original_order}
ExecStart=/usr/bin/rm /var/lib/arch-ops-server/boot_order_restore
ExecStart=/usr/bin/systemctl disable arch-ops-boot-restore.service
RemainAfterExit=no

[Install]
WantedBy=basic.target
"""

    _STATE_FILE.parent.mkdir(parents=True, exist_ok=True)
    _STATE_FILE.write_text(original_order)

    try:
        _SERVICE_FILE.parent.mkdir(parents=True, exist_ok=True)
        _SERVICE_FILE.write_text(service_content)
    except OSError as e:
        return create_error_response("WriteError", f"Failed to write service file: {str(e)}")

    exit_code, _, stderr = await run_command(
        ["systemctl", "daemon-reload"],
        timeout=5,
        check=False,
    )
    if exit_code != 0:
        return create_error_response("SystemdError", f"daemon-reload failed: {stderr}")

    exit_code, _, stderr = await run_command(
        ["systemctl", "enable", "arch-ops-boot-restore.service"],
        timeout=5,
        check=False,
    )
    if exit_code != 0:
        return create_error_response("SystemdError", f"enable failed: {stderr}")

    return {"restore_service_installed": True, "original_order": original_order}


async def manage_boot(
    action: str,
    order: Optional[str] = None,
    device: Optional[str] = None,
    reboot: bool = False,
) -> Dict[str, Any]:
    """
    Manage Raspberry Pi bootloader configuration.

    Actions:
      status         — read current BOOT_ORDER, decode it, detect available boot devices
      set_boot_order — permanently change BOOT_ORDER (use preset name or raw hex)
      next_boot      — set next boot device as one-time, reboot
    """
    if not IS_ARCH:
        return create_error_response(
            "NotSupported",
            f"manage_boot(action='{action}') requires Arch Linux running on physical RPi hardware.",
        )

    if not check_command_exists("rpi-eeprom-config"):
        return create_error_response(
            "NotSupported",
            "rpi-eeprom-config is not installed. Install with: sudo pacman -S rpi-eeprom",
        )

    if action == "status":
        logger.info("Reading bootloader configuration")
        config = await _get_eeprom_config()
        if not config:
            return create_error_response("CommandError", "Failed to read EEPROM configuration")

        boot_order = config.get("BOOT_ORDER", "unknown")
        sequence = _decode_boot_order(boot_order)
        devices = await _detect_boot_devices()
        has_restore = _STATE_FILE.exists()

        return {
            "boot_order": boot_order,
            "sequence": sequence,
            "available_devices": devices,
            "has_pending_restore": has_restore,
            "config_keys": list(config.keys()),
        }

    elif action == "set_boot_order":
        if not order:
            return create_error_response("MissingArgument", "order parameter is required")

        if order in _ORDER_PRESETS:
            boot_order = _ORDER_PRESETS[order]
        else:
            if not re.match(r"^0x[0-9a-fA-F]+$", order):
                return create_error_response(
                    "InvalidValue",
                    f"Invalid boot order: '{order}'. Use a preset ({', '.join(_ORDER_PRESETS.keys())}) or raw hex like 0xf416.",
                )
            boot_order = order

        logger.info(f"Setting BOOT_ORDER to {boot_order}")
        result = await _run_eeprom_update("BOOT_ORDER", boot_order)
        if "error" in result:
            return result

        sequence = _decode_boot_order(boot_order)
        return {
            "set": True,
            "boot_order": boot_order,
            "sequence": sequence,
            "message": f"BOOT_ORDER set to {boot_order}. Reboot to apply.",
        }

    elif action == "next_boot":
        if not device:
            return create_error_response("MissingArgument", "device parameter is required (sd, nvme, usb)")

        device = device.lower()
        if device not in ("sd", "nvme", "usb"):
            return create_error_response(
                "InvalidValue",
                f"Invalid device: '{device}'. Use: sd, nvme, usb.",
            )

        config = await _get_eeprom_config()
        if not config:
            return create_error_response("CommandError", "Failed to read EEPROM configuration")

        original_order = config.get("BOOT_ORDER", "0xf41")
        logger.info(f"Saving original BOOT_ORDER: {original_order}")

        device_nibble = {"sd": "1", "nvme": "6", "usb": "4"}[device]
        temp_order = f"0xf{device_nibble}"

        result = await _run_eeprom_update("BOOT_ORDER", temp_order)
        if "error" in result:
            return result

        result = await _install_restore_service(original_order)
        if "error" in result:
            return result

        if not reboot:
            return {
                "temp_order_set": True,
                "temp_order": temp_order,
                "original_order_saved": original_order,
                "restore_service_installed": True,
                "message": f"Next boot set to {device.upper()} ({temp_order}). Reboot manually to apply.",
            }

        logger.warning(f"Rebooting to {device.upper()} device")
        exit_code, _, stderr = await run_command(
            ["reboot"],
            timeout=5,
            check=False,
        )
        if exit_code != 0:
            return create_error_response("CommandError", f"Reboot failed: {stderr}")

        return {
            "rebooting": True,
            "device": device,
            "temp_order": temp_order,
            "message": f"Rebooting to {device.upper()} device. Original BOOT_ORDER ({original_order}) will be restored automatically.",
        }

    else:
        return create_error_response(
            "InvalidAction",
            f"Unknown action: '{action}'. Valid actions: status, set_boot_order, next_boot.",
        )

File: src/tools/boot.rs
use crate::error::ToolError;
use crate::platform::is_arch_linux;
use crate::protocol::Tool;
use async_trait::async_trait;
use serde_json::Value;
use std::collections::HashMap;

use super::ToolHandler;

pub struct ManageBoot;

#[async_trait]
impl ToolHandler for ManageBoot {
    fn info(&self) -> Tool {
        Tool {
            name: "manage_boot".into(),
            description: "Manage Raspberry Pi bootloader (BOOT_ORDER, one-time boot via tryboot). Supports RPi5 with rpi-eeprom-config.".into(),
            input_schema: serde_json::json!({
                "type": "object",
                "properties": {
                    "action": {
                        "type": "string",
                        "enum": ["status", "set_boot_order", "next_boot"],
                        "description": "Boot management action"
                    },
                    "order": {
                        "type": "string",
                        "description": "Boot order preset or raw hex. Presets: sd_first, nvme_first, usb_first, sd_nvme, nvme_sd, sd_only, nvme_only, usb_only"
                    },
                    "device": {
                        "type": "string",
                        "enum": ["sd", "nvme", "usb"],
                        "description": "Device for next_boot"
                    },
                    "reboot": { "type": "boolean", "default": false, "description": "Reboot after setting" }
                },
                "required": ["action"]
            }),
        }
    }

    async fn call(&self, args: HashMap<String, Value>) -> Result<Value, ToolError> {
        if !is_arch_linux() {
            return Err(ToolError::platform_not_arch());
        }

        let action = args
            .get("action")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::invalid_argument("Missing: action"))?;

        match action {
            "status" => boot_status().await,
            "set_boot_order" => {
                let order = args
                    .get("order")
                    .and_then(|v| v.as_str())
                    .ok_or_else(|| ToolError::invalid_argument("Missing: order"))?;
                let reboot = args
                    .get("reboot")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(false);
                set_boot_order(order, reboot).await
            }
            "next_boot" => {
                let device = args
                    .get("device")
                    .and_then(|v| v.as_str())
                    .ok_or_else(|| ToolError::invalid_argument("Missing: device"))?;
                let reboot = args
                    .get("reboot")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(false);
                next_boot(device, reboot).await
            }
            _ => Err(ToolError::invalid_argument("Unknown action")),
        }
    }
}

fn boot_order_preset(name: &str) -> Option<&'static str> {
    match name {
        "sd_first" => Some("0xf461"),
        "nvme_first" => Some("0xf462"),
        "usb_first" => Some("0xf463"),
        "sd_nvme" => Some("0xf14"),
        "nvme_sd" => Some("0xf24"),
        "sd_only" => Some("0x1"),
        "nvme_only" => Some("0x2"),
        "usb_only" => Some("0x3"),
        _ => None,
    }
}

fn decode_boot_order(hex: &str) -> String {
    let clean = hex.trim_start_matches("0x").trim();
    let devices: Vec<&str> = clean
        .chars()
        .filter_map(|c| match c {
            '1' => Some("SD"),
            '2' => Some("NVMe"),
            '3' => Some("USB"),
            '4' => Some("SD (alt)"),
            '5' => Some("NVMe (alt)"),
            '6' => Some("USB (alt)"),
            'e' => Some("EEPROM"),
            'f' => Some("Restart"),
            '0' => None,
            _ => Some(&c.to_string()),
        })
        .collect();

    if devices.is_empty() {
        hex.to_string()
    } else {
        devices.join(" → ")
    }
}

async fn read_bootloader_config() -> Result<String, ToolError> {
    let result = crate::command::run("rpi-eeprom-config", &[]).await?;
    Ok(result.stdout)
}

async fn boot_status() -> Result<Value, ToolError> {
    let config = read_bootloader_config().await?;
    let boot_order_line = config.lines().find(|l| l.starts_with("BOOT_ORDER"));
    let boot_order = boot_order_line
        .and_then(|l| l.split('=').nth(1))
        .map(|v| v.trim())
        .unwrap_or("unknown");

    let eeprom_version = config
        .lines()
        .find(|l| l.starts_with("FIRMWARE_VERSION") || l.starts_with("BOOTLOADER_VERSION"));

    Ok(serde_json::json!({
        "boot_order": boot_order,
        "boot_order_decoded": decode_boot_order(boot_order),
        "eeprom_version": eeprom_version,
        "config": config
    }))
}

async fn set_boot_order(order: &str, reboot: bool) -> Result<Value, ToolError> {
    let hex_value = boot_order_preset(order).unwrap_or(order);

    let current_config = read_bootloader_config().await?;
    let updated_config = current_config
        .lines()
        .map(|l| {
            if l.starts_with("BOOT_ORDER") {
                format!("BOOT_ORDER={}", hex_value.trim_start_matches("0x"))
            } else {
                l.to_string()
            }
        })
        .collect::<Vec<_>>()
        .join("\n");

    // Write via temp file + rpi-eeprom-config --apply
    let temp_path = "/tmp/boot-eeprom-config.txt";
    std::fs::write(temp_path, &updated_config)
        .map_err(|e| ToolError::internal(&format!("Cannot write temp config: {}", e)))?;

    let result = crate::command::run("sudo", &["rpi-eeprom-config", "--apply", temp_path]).await?;
    let _ = std::fs::remove_file(temp_path);

    if reboot {
        let _ = crate::command::run("sudo", &["reboot"]).await;
    }

    Ok(serde_json::json!({
        "action": "set_boot_order",
        "order": order,
        "hex": hex_value,
        "output": result.stdout,
        "exit_code": result.exit_code,
        "reboot": reboot
    }))
}

async fn next_boot(device: &str, reboot: bool) -> Result<Value, ToolError> {
    let current_config = read_bootloader_config().await?;
    let original_order = current_config
        .lines()
        .find(|l| l.starts_with("BOOT_ORDER"))
        .map(|l| l.to_string())
        .unwrap_or_default();

    let tryboot_order = match device {
        "sd" => "0x1",
        "nvme" => "0x2",
        "usb" => "0x3",
        _ => return Err(ToolError::invalid_argument("Unknown device")),
    };

    let tryboot_config = current_config
        .lines()
        .map(|l| {
            if l.starts_with("BOOT_ORDER") {
                format!("BOOT_ORDER={}", tryboot_order.trim_start_matches("0x"))
            } else {
                l.to_string()
            }
        })
        .collect::<Vec<_>>()
        .join("\n");

    let temp_path = "/tmp/boot-eeprom-config.txt";
    std::fs::write(temp_path, &tryboot_config)
        .map_err(|e| ToolError::internal(&format!("Cannot write temp config: {}", e)))?;

    let result = crate::command::run("sudo", &["rpi-eeprom-config", "--apply", temp_path]).await?;
    let _ = std::fs::remove_file(temp_path);

    if reboot {
        let _ = crate::command::run("sudo", &["reboot"]).await;
    }

    Ok(serde_json::json!({
        "action": "next_boot",
        "device": device,
        "original_boot_order": original_order,
        "temporary_boot_order": tryboot_order,
        "note": "BOOT_ORDER will be restored after successful boot via systemd oneshot",
        "output": result.stdout,
        "reboot": reboot
    }))
}
